concordium-consensus/src/Concordium/KonsensusV1/LeaderElection.hs

```diff
@@ -0,0 +1,95 @@
+{-# LANGUAGE DataKinds #-}
+
+module Concordium.KonsensusV1.LeaderElection (
+    -- * Leader election
+    getLeader,
+
+    -- * Seed state
+    updateSeedStateForBlock,
+    updateSeedStateForEpoch,
+) where
+
+import Data.Serialize
+
+import qualified Concordium.Crypto.SHA256 as Hash
+import Concordium.Types
+import Concordium.Types.HashableTo
+import Concordium.Types.SeedState
+
+import Concordium.Crypto.VRF (proofToHash)
+import Concordium.GlobalState.BakerInfo
+import Concordium.KonsensusV1.Types
+
+-- |Compute the leader for a given round, given the set of bakers and leadership election nonce
+-- for the epoch. The leader is computed as follows:
+--
+--  * Compute the SHA256 hash of the nonce and round number.
+--  * Treating this as an big-endian unsigned integer, compute the remainder when divided by the
+--    total stake among all the bakers, @targetVal@.
+--  * Find the least index @i@ such that the sum of stakes of bakers with indices less than or equal
+--    to @i@ is greater than @targetVal@. The baker with this index is the leader.
+--
+-- This process has a bias towards bakers with smaller indexes. However, the chance of this bias
+-- affecting the result of any given election is less than 1 in 10^57, and so is effectively
```

Comment:
Where does the bias come from and why 10^57?

Reply:
The hash function generates values in the range `[0 .. 2^256 - 1]`. We want to sample a value in the range `[0 .. totalStake - 1]`, where `totalStake` is going to be at most `2^64`. If we sampled from a range `[0 .. k * totalStake - 1]` (for integer `k`), then taking the remainder modulo `totalStake` gives a uniformly distributed output (assuming the input is uniform). If we select `k` to be maximal such that `k * totalStake - 1 <= 2^256 -1` then we can see that there are at most 2^64 values in the range `[0 .. 2^256 - 1]` that are not in the range `[0 .. k * totalStake - 1]`. If we want to eliminate bias completely, then we can say, if we get such a value then we resample. But the chance of us getting such a value is (2^64 / 2^256) = 1.593... * 10^-58. Hence less than 1 in 10^57.